### ISSUE TITLE
Load with user-defined types

### ISSUE THREAD

**Issue by chrisaycock** [empirical-soft/empirical-lang#23]
TL;DR From the roadmap (#1), we want to supply a user-defined type to `load()`:

```
load{MySwankyType}("some_file.csv")
```

This also removes the dollar sign (`$`) from the user's sight.

----

## Goal

Empirical can load a CSV file into a DataFrame while statically inferring the type. This is an excellent feature for quickly working in a REPL.

However, the current mechanism hard-codes the logic to infer and then load. This means the user can't override the inferred type. Worse, it means that the sample file must be available at compile-time, which is impossible in many production settings.

A better method is to have **type parameters** that the user can supply for function calls and data definitions. Consider the canonical loader:

```
let trades = load("trades.csv")
```

The above is equivalent to:

```
data Trade = Provider{"trades.csv"}
let trades = load{Trade}("trades.csv")
```

The above functions can be expressed as:

```
# type provider (ala F#)
data Provider{filename: String} = compile(_infer(filename))

# templated version of load that takes a type
func load{T: Kind}(filename: String) = coerce(!T, _load(T, filename))

# macro that overloads load
macro load(filename: String) = load{Provider{filename}}(filename)
```

The `_infer` and `_load` are VVM functions, while `compile` and `coerce` are directives to the Empirical compiler.

## Roadmap

Empirical needs major upgrades to perform `load` in a general way.

* Function definitions via expression syntax (ala Kotlin) that are inlined:

```
func inc(x) = x + 1

inc(7)    # inlines to 7 + 1
```

* Data definitions via expression syntax:

```
data SizeT = Int64
data Person = {name: String, age: Int64}
```

* Templated functions:

```
func add{T: Kind}(x: T, y: T):
  return x + y
end

add{Int64}(3, 4)
```

* Templated datatypes:

```
data Person{AgeType: Kind}:
  name: String,
  age: AgeType
end

Person{Int64}("a", 13)
!Person{Float64}(["a", "b"], [1.3, 1.4])
```

* Macros that look like functions, but expand on invocation at compile time:

```
macro inc(x: Int64) = x + 1    # parameter must be comptime
```

* Evaluate a string into the AST (ala D's `mixin`) for statements and expressions:

```
let y = compile("x + 1")    # equivalent to: let y = x + 1

compile("func inc(x: Int64): return x + 1 end")
```

* Type coercion to reinterpret data:

```
coerce(Int64, "Hello World")    # not safe, but sometimes needed for VVM functions
```

* Generic types (*bonus*):

```
func add(x, y): return x + y end
```
